src/dotnet/ToolManifest/ToolManifestEditor.cs

```diff
@@ -217,6 +218,44 @@ private class SerializableLocalToolsManifest
             [JsonProperty(Required = Required.Always)]
             // The dictionary's key is the package id
             public Dictionary<string, SerializableLocalToolSinglePackage> tools { get; set; }
+
+            public SerializableLocalToolsManifest With(Dictionary<string, SerializableLocalToolSinglePackage> newTools)
+            {
+                return new SerializableLocalToolsManifest
```

Comment:
Do we need to return a new SerializableLocalToolsManifest rather than just setting the `tools` property and returning `this`?

Reply:
I'm not sure I understand that.  Typically "With" methods are just manipulating the current instance and returning `this` to enable chaining of setting various properties prior to calling some "action" method.  As a caller to a `With*` method, I would not expect a new instance.

Example:
https://github.com/dotnet/cli/blob/95c6eff6daa1a69f29c42b2d405400ad44bdec91/test/Microsoft.DotNet.Tools.Tests.Utilities/Extensions/TestCommandExtensions.cs#L16-L21

And in non-test code (although without the "With" prefix):

https://github.com/dotnet/cli/blob/95c6eff6daa1a69f29c42b2d405400ad44bdec91/test/Microsoft.DotNet.Tools.Tests.Utilities/Extensions/TestCommandExtensions.cs#L16-L21

Reply:
It is the opposite, typically "With-er" is convention to "edit" immutable object as in this proposal https://github.com/dotnet/csharplang/blob/master/proposals/records.md and it is unfortunate common aspnet builder pattern manipulate the same object.

However, i am fine changing the method name from `With` to `NewCopyWith` to distinguish it

Reply:
Renamed to NewCopyWithTools

Reply:
Sorry, I didn't see your previous comment.  I'm much more familiar with the use of "WithXYZ" methods primarily used in builders (not just C# pattern too) and hadn't considered a more generic `With` method that would take all properties as arguments to create a copy with optional replacement data, which would only really be nice (syntactically-speaking) with caller-receiver parameters.

The more I think about the intention of this method, the more I am inclined to just remove this method entirely and simply set the `Tools` property where this is being called instead (to leave a better `With` method for the future when C# syntax allows).

```
deserializedManifest.Tools = deserializedManifest.Tools
                .Where(pair => !pair.Key.Equals(packageId.ToString(), StringComparison.Ordinal))
                .ToDictionary(pair => pair.Key, pair => pair.Value);

 _fileSystem.File.WriteAllText(
                fromFilePath.Value,
                JsonConvert.SerializeObject(deserializedManifest, Formatting.Indented));
```